django/contrib/admin/static/admin/js/nav_sidebar.js

```diff
@@ -0,0 +1,22 @@
+'use strict';
+{
+    const toggleNavSidebar = document.getElementById('toggle-nav-sidebar');
+    if (toggleNavSidebar !== null) {
+        const main = document.getElementById('main');
+        let navSidebarIsOpen = localStorage.getItem('django.admin.navSidebarIsOpen');
+        if (navSidebarIsOpen === null) {
+            navSidebarIsOpen = 'true';
+        }
+        main.classList.toggle('shifted', navSidebarIsOpen === 'true');
+
+        toggleNavSidebar.addEventListener('click', function() {
+            if (navSidebarIsOpen == 'true') {
+                navSidebarIsOpen = 'false';
+            } else {
+                navSidebarIsOpen = 'true';
+            }
+            localStorage.setItem('django.admin.navSidebarIsOpen', navSidebarIsOpen);
+            main.classList.toggle('shifted');
+        });
+    }
+}
```

Comment:
I think a selenium test is needed for this.

Reply:
I tried to add some selenium tests, but ran into a problem. It seems like even if the navbar is closed and not visible to the user, `is_displayed()` still returns `True`. It seems to be this issue: https://github.com/SeleniumHQ/selenium/issues/4689

I think we might be able to get around this by adding `display: none` in the CSS, it doesn't seem to break anything, but I'm wary of changing the CSS at this point to get a test to work.

Reply:
Hmmm. 🤔

How about a QUnit test in `js_tests` then. We can assume local storage works. 
Just testing that the nav bar toggles, and sets to and reads from local storage correctly.

(Obviously it does, but at least we have a regression test then.) 

?

Reply:
I had a look at this, but I'm struggling with QUnit. Maybe I'm missing something. I added a fixture into tests.html and the relevant scripts:

```html
<script type="text/html" id="sidebar">
  <div class="sticky toggle-nav-sidebar" id="toggle-nav-sidebar"></div>
  <nav class="sticky" id="nav-sidebar"></nav>
  <div id="main"></div>
</script>

<script src="../django/contrib/admin/static/admin/js/nav_sidebar.js" data-cover></script>
<script src="./admin/NavSidebar.test.js"></script>
```

And a simple test:
```javascript
/* global QUnit, NavSidebar */
'use strict';

QUnit.module('admin.NavSidebar', {
    beforeEach: function() {
        const $ = django.jQuery;
        $('#qunit-fixture').append($('#sidebar').text());
    }
});

QUnit.test('sets local storage', function(assert) {
    const $ = django.jQuery;
    assert.equal(localStorage.getItem('django.admin.navSidebarIsOpen'), 'true');
    $('#toggle-nav-sidebar').click();
    assert.equal(localStorage.getItem('django.admin.navSidebarIsOpen'), 'false');
});
```

And I get this:
```
admin.NavSidebar - sets local storage...ERROR
>> Message: null
>> Actual: null
>> Expected: "true"
>> at Object.<anonymous> (file:///.../django/js_tests/admin/NavSidebar.test.js:13:12)

>> Message: null
>> Actual: null
>> Expected: "false"
>> at Object.<anonymous> (file:///.../django/js_tests/admin/NavSidebar.test.js:15:12)
```

I'm not exactly sure why. I get the feeling the script is running before the fixture is in, so the code doesn't run because it checks first if the element is on the page, but not having any idea how QUnit works, I don't know.

Reply:
Grrr. OK, the issue with the QUnit test is (I think) that the event listeners and such are not attached to the the `#sidebar` HTML when the page is loaded (when the `nav_sidebar.js` is loaded) because they're not part of the DOM at that point. 

When the test case is run, they are inserted into the DOM but we don't have an `init()` type function that we can call to (re)attach the even handlers. (cp the Select2 tests, which call Select2.init() with the added HTML)

So this...

```javascript

'use strict';

QUnit.module('admin.NavSidebar', {
    beforeEach: function() {
        const $ = django.jQuery;
        $('#qunit-fixture').append($('#sidebar').text());
        this.toggle = $("#toggle-nav-sidebar");
        this.sidebar = $("#nav-sidebar");
        this.main = $('#main');

    }
});

QUnit.test('sets local storage', function(assert) {
    const $ = django.jQuery;
    assert.ok(!this.main.hasClass('shifted'));
    $('#toggle-nav-sidebar').click();
    assert.ok(this.main.hasClass('shifted'));
});
```

Fails at the last assert, because the event handler is missing. 

I think the best bet is a selenium test that just checks the correct classes are toggled, and uses (I guess) `execute_script()` to check local storage is set. 

I don't think we need more that that. (That the CSS worked to adjust what's visible isn't something I want to use selenium to assert here... 😳)

Do you think you can add that @knyghty?

Reply:
Done, seems to be working okay.